### PR TITLE
Update example on extending data preprocessing

### LINKED ISSUE #1257: Turning of the data preprocessing step causes algorithms to crash
## Describe the bug
I have updated Auto-Sklearn to the latest version to be able to completely turn off any preprocessing. I have followed the [example](https://automl.github.io/auto-sklearn/master/examples/80_extending/example_extending_data_preprocessor.html#sphx-glr-examples-80-extending-example-extending-data-preprocessor-py) on the webpage and, it seems, turning off the preprocessing introduces crashes.

The provided example on the webpage seems to suffer the same issue as MyDummyClassifier is returned in the ensemble.

## To Reproduce
    # Imports and code for the NoPreprocessing class, as shown in the example

    X, y = load_breast_cancer(return_X_y=True)

    automl = AutoSklearnClassifier(
        time_left_for_this_task=30,
        include={
            'data_preprocessor': ['NoPreprocessing'],
            'classifier': ['random_forest']
        },
        initial_configurations_via_metalearning=0
    ).fit(X, y)

    print(automl.sprint_statistics())

## Expected behavior
By removing the `data_preprocessor`, I obtain the expected behaviour with the expected output.

    # Imports

    X, y = load_breast_cancer(return_X_y=True)

    automl = AutoSklearnClassifier(
        time_left_for_this_task=30,
        include={
            'classifier': ['random_forest']
        },
        initial_configurations_via_metalearning=0
    ).fit(X, y)

    print(automl.sprint_statistics())

Output:

    auto-sklearn results:
      Dataset name: 5f06c41c-1fbb-11ec-b878-acde48001122
      Metric: accuracy
      Best validation score: 0.957447
      Number of target algorithm runs: 10
      Number of successful target algorithm runs: 7
      Number of crashed target algorithm runs: 2
      Number of target algorithms that exceeded the time limit: 1
      Number of target algorithms that exceeded the memory limit: 0

## Actual behavior, stacktrace or logfile
Output, when `data_preprocessor = ['NoPreprocessing']` is used:

    auto-sklearn results:
      Dataset name: 2df246e0-1fba-11ec-b827-acde48001122
      Metric: accuracy
      Number of target algorithm runs: 22
      Number of successful target algorithm runs: 0
      Number of crashed target algorithm runs: 22
      Number of target algorithms that exceeded the time limit: 0
      Number of target algorithms that exceeded the memory limit: 0

## Environment and installation

* macOS Catalina
* Python's `venv` module
* Python 3.9.7
* auto-sklearn 0.14.0

### PR DESCRIPTION
**Pull request by eddiebergman**
This PR fixes a broken example on extending autosklearn with a `NoPreprocessing` step for data preprocessing.
It also updates the docs to distinguish this.

Closes #1257

### CODE REVIEW

**Comment by mdbecker** [comment]
FYI I tried out the example code from this PR on my dataset (which has no NaNs, every variable is a float between 0.0 and 1.0) and it's crashing.

I see this in the logs:
```
Error reading configurations:Trying to set illegal value 'feature_type' (type '<class 'str'>') for hyperparameter 'data_preprocessor:__choice__, Type: Categorical, Choices: {NoPreprocessing}, Default: NoPreprocessing' (default-value has type '<class 'str'>').
```

And here is the output of `sprint_statistics`:
```
auto-sklearn results:
  Dataset name: 4743dfe0-38e2-11ec-86b8-0242ac110002
  Metric: roc_auc
  Number of target algorithm runs: 92
  Number of successful target algorithm runs: 0
  Number of crashed target algorithm runs: 92
  Number of target algorithms that exceeded the time limit: 0
  Number of target algorithms that exceeded the memory limit: 0
```

**Comment by mdbecker** [comment]
I tried the example and it works as is, but if you change `n_jobs` it fails with `Number of crashed target algorithm runs: 5`. I assume this has something to do with Dask but don't know enough to be sure. Here is a minimum example to reproduce the failure assuming you've done everything else in the example code:
```
clf = AutoSklearnClassifier(
    time_left_for_this_task=120,
    include={
        'data_preprocessor': ['NoPreprocessing']
    },
    # Bellow two flags are provided to speed up calculations
    # Not recommended for a real implementation
    initial_configurations_via_metalearning=0,
    smac_scenario_args={'runcount_limit': 5},
    n_jobs=7
)
clf.fit(X_train, y_train)
print(clf.sprint_statistics())
```

**Comment by eddiebergman** [comment]
> I tried the example and it works as is, but if you change `n_jobs` it fails with `Number of crashed target algorithm runs: 5`. I assume this has something to do with Dask but don't know enough to be sure. Here is a minimum example to reproduce the failure assuming you've done everything else in the example code:
> 
> ```
> clf = AutoSklearnClassifier(
>     time_left_for_this_task=120,
>     include={
>         'data_preprocessor': ['NoPreprocessing']
>     },
>     # Bellow two flags are provided to speed up calculations
>     # Not recommended for a real implementation
>     initial_configurations_via_metalearning=0,
>     smac_scenario_args={'runcount_limit': 5},
>     n_jobs=7
> )
> clf.fit(X_train, y_train)
> print(clf.sprint_statistics())
> ```

Hi @mdbecker,

Thanks for reporting this, I'll have a look into it but I'm not sure why that would be.

File: examples/80_extending/example_extending_data_preprocessor.py [hunk 1]
```diff
@@ -23,22 +23,20 @@ class NoPreprocessing(AutoSklearnPreprocessingAlgorithm):
 
     def __init__(self, **kwargs):
         """ This preprocessors does not change the data """
-        self.preprocessor = None
+        # Some internal checks makes sure parameters are set
```
Comment (mfeurer) [inline examples/80_extending/example_extending_data_preprocessor.py:26]: What do you mean by `some internal checks`?

File: examples/80_extending/example_extending_data_preprocessor.py [hunk 2]
```diff
@@ -82,6 +79,9 @@ def get_hyperparameter_search_space(dataset_properties=None):
 )
 clf.fit(X_train, y_train)
 
+assert len(clf.get_models_with_weights()) > 0  # Just to make sure models were found
```
Comment (mfeurer) [inline examples/80_extending/example_extending_data_preprocessor.py:82]: Could you please extend the comment so that it is more standalone?

File: examples/80_extending/example_extending_data_preprocessor.py (hunk 1 above)
Comment (eddiebergman) [reply examples/80_extending/example_extending_data_preprocessor.py:26]: I'm not entirely sure where it was but there is a check somewhere in the pipeline that checks that certain attributes are set on the object from the `**kwargs` hence the snippet right below this
```python
for key, val in kwargs.items():
    setattr(self, key, val)
```

File: examples/80_extending/example_extending_data_preprocessor.py (hunk 2 above)
Comment (eddiebergman) [reply examples/80_extending/example_extending_data_preprocessor.py:82]: Done

**Comment by eddiebergman** [comment]
Hi @mdbecker,

I couldn't reproduce the failure, I've put my entire code snippet below which could hopefully help diagnose any issues.

```python
import autosklearn
from autosklearn.classification import AutoSklearnClassifier
from autosklearn.pipeline.components.base import AutoSklearnPreprocessingAlgorithm
from sklearn.datasets import load_breast_cancer
import sklearn.metrics
from autosklearn.pipeline.constants import SPARSE, DENSE, UNSIGNED_DATA, INPUT
from sklearn.model_selection import train_test_split
from ConfigSpace.configuration_space import ConfigurationSpace

X, y = load_breast_cancer(return_X_y=True)


class NoPreprocessing(AutoSklearnPreprocessingAlgorithm):

    def __init__(self, **kwargs):
        """This preprocessors does not change the data"""
        # Some internal checks makes sure parameters are set
        for key, val in kwargs.items():
            setattr(self, key, val)

    def fit(self, X, Y=None):
        return self

    def transform(self, X):
        return X

    @staticmethod
    def get_properties(dataset_properties=None):
        return {
            "shortname": "NoPreprocessing",
            "name": "NoPreprocessing",
            "handles_regression": True,
            "handles_classification": True,
            "handles_multiclass": True,
            "handles_multilabel": True,
            "handles_multioutput": True,
            "is_deterministic": True,
            "input": (SPARSE, DENSE, UNSIGNED_DATA),
            "output": (INPUT,),
        }

    @staticmethod
    def get_hyperparameter_search_space(dataset_properties=None):
        return ConfigurationSpace()  # Return an empty configuration as there is None


# Add NoPreprocessing component to auto-sklearn.
autosklearn.pipeline.components.data_preprocessing.add_preprocessor(NoPreprocessing)


if __name__ == "__main__":

    clf = AutoSklearnClassifier(
        time_left_for_this_task=120,
        include={
            'data_preprocessor': ['NoPreprocessing']
        },
        # Bellow two flags are provided to speed up calculations
        # Not recommended for a real implementation
        initial_configurations_via_metalearning=0,
        smac_scenario_args={'runcount_limit': 5},
        n_jobs=7
    )
    clf.fit(X, y)
    print(clf.sprint_statistics())
```



Output:
```bash
[WARNING] [2021-11-03 11:30:21,982:Client-AutoML(1):0bb3609f-3c91-11ec-a202-ec7949506548] Capping the per_run_time_limit to 59.0 to have time for a least 2 models in each process.
auto-sklearn results:
  Dataset name: 0bb3609f-3c91-11ec-a202-ec7949506548
  Metric: accuracy
  Best validation score: 0.957447
  Number of target algorithm runs: 5
  Number of successful target algorithm runs: 4
  Number of crashed target algorithm runs: 1
  Number of target algorithms that exceeded the time limit: 0
  Number of target algorithms that exceeded the memory limit: 0
```
